class.photon.php

```diff
@@ -646,9 +650,86 @@ public function filter_srcset_array( $sources, $size_array, $image_src, $image_m
 			$sources[ $i ]['url'] = jetpack_photon_url( $url, $args );
 		}
 
+		/**
+		 * At this point, $sources is the original srcset with Photonized URLs.
+		 * Now, we're going to construct additional sizes based on percent of the original.
+		 * This will reduce the gap between the largest defined size and the original image.
+		 */
+
+		/**
+		 * Filter the multiplier Photon uses to create new srcset items.
+		 * Return false to short-circuit and bypass auto-generation.
+		 *
+		 * @module photon
+		 *
+		 * @since 4.1.0
+		 *
+		 * @param array|bool $multipliers Array of multipliers to use or false to bypass.
+		 */
+		$multipliers = apply_filters( 'jetpack_photon_srcset_multipliers', array(
+			'content_width' => array( 2, 3 ),
+			'full_size'     => array( 0.8, 0.6, 0.4 ),
+			 ) );
+		// Allow short-circuiting
+		if ( ! is_array( $multipliers ) ){
+			return $sources;
+		}
+
+		// If the meta isn't complete, something likely broke when uploading the original.
+		if ( isset( $image_meta['width'] ) && isset( $image_meta['file'] ) ) {
+			$url = trailingslashit( $upload_dir['baseurl'] ) . $image_meta['file'];
+
+			$currentwidths = array_keys( $sources );
+
+			if ( is_array( $multipliers['content_width'] ) && $content_width = Jetpack::get_content_width() ) {
+				$multipliers = $multipliers['content_width'];
+				$base        = $content_width;
+			} elseif ( is_array( $multipliers['full_size'] ) ) { // If no $content_width is set, let's make variations of the full-sized image.
+				$multipliers = $multipliers['full_size'];
+				$base        = $image_meta['width'];
+			} else { // In case someone managled the filter, let's return instead of mangling things.
+				return $sources;
+			}
+
+			foreach ( $multipliers as $multiplier ){
+				$usewidth = true;
+				$newwidth = round( $base * $multiplier );
+				foreach ( $currentwidths as $currentwidth ){
+					// If a new width would be within 100 pixes of an existing one, skip.
+					if ( abs( $currentwidth - $newwidth ) < 50 ) {
+						$usewidth = false;
+						break;
```

Comment:
It's an uncommon syntax that may harm readability, but there's a case for [specifying a numeric argument for `continue`](http://php.net/manual/en/control-structures.continue.php) here:

```php
foreach ( $multipliers as $multiplier ) {
	$newwidth = round( $base * $multiplier );
	foreach ( $currentwidths as $currentwidth ) {
		// If a new width would be within 100 pixes of an existing one, skip.
		if ( abs( $currentwidth - $newwidth ) < 50 ) {
			continue 2;
		}
	}

	$newsources[ $newwidth ] = array(
		'url'         => jetpack_photon_url( $url, array( 'w' => $newwidth ) ),
		'descriptor'  => 'w',
		'value'       => $newwidth,
	);
} // foreach ( $multipliers as $multiplier )
```